includes/collection/class-followers.php

```diff
@@ -0,0 +1,383 @@
+				},
+			)
+		);
+
+		register_term_meta(
+			self::TAXONOMY,
+			'username',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'sanitize_callback' => function( $value ) {
+					return sanitize_user( $value, true );
+				},
+			)
+		);
+
+		register_term_meta(
+			self::TAXONOMY,
+			'avatar',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'sanitize_callback' => function( $value ) {
+					if ( filter_var( $value, FILTER_VALIDATE_URL ) === false ) {
+						return '';
+					}
+
+					return esc_url_raw( $value );
+				},
+			)
+		);
+
+		register_term_meta(
+			self::TAXONOMY,
+			'inbox',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'sanitize_callback' => function( $value ) {
+					if ( filter_var( $value, FILTER_VALIDATE_URL ) === false ) {
+						throw new Exception( '"inbox" has to be a valid URL' );
+					}
+
+					return esc_url_raw( $value );
+				},
+			)
+		);
+
+		register_term_meta(
+			self::TAXONOMY,
+			'shared_inbox',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'sanitize_callback' => function( $value ) {
+					if ( filter_var( $value, FILTER_VALIDATE_URL ) === false ) {
+						return null;
+					}
+
+					return esc_url_raw( $value );
+				},
+			)
+		);
+
+		register_term_meta(
+			self::TAXONOMY,
+			'updated_at',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'sanitize_callback' => function( $value ) {
+					if ( ! is_numeric( $value ) && (int) $value !== $value ) {
+						$value = strtotime( 'now' );
+					}
+
+					return $value;
+				},
+			)
+		);
+
+		do_action( 'activitypub_after_register_taxonomy' );
+	}
+
+	/**
+	 * Handle the "Follow" Request
+	 *
+	 * @param array $object    The JSON "Follow" Activity
+	 * @param int   $user_id The ID of the ID of the WordPress User
+	 *
+	 * @return void
+	 */
+	public static function handle_follow_request( $object, $user_id ) {
+		// save follower
+		$follower = self::add_follower( $user_id, $object['actor'] );
+
+		do_action( 'activitypub_followers_post_follow', $object['actor'], $object, $user_id, $follower );
+	}
+
+	/**
+	 * Handles "Unfollow" requests
+	 *
+	 * @param array $object  The JSON "Undo" Activity
+	 * @param int   $user_id The ID of the ID of the WordPress User
+	 */
+	public static function handle_undo_request( $object, $user_id ) {
+		if (
+			isset( $object['object'] ) &&
+			isset( $object['object']['type'] ) &&
+			'Follow' === $object['object']['type']
+		) {
+			self::remove_follower( $user_id, $object['actor'] );
+		}
+	}
+
+	/**
+	 * Add a new Follower
+	 *
+	 * @param int    $user_id The ID of the WordPress User
+	 * @param string $actor   The Actor URL
+	 *
+	 * @return array|WP_Error The Follower (WP_Term array) or an WP_Error
+	 */
+	public static function add_follower( $user_id, $actor ) {
+		$meta = get_remote_metadata_by_actor( $actor );
```

Comment:
It is important to note that because of this requirement, it could be that a follower might not be converted. For example if their Mastodon server is temporarily down.

I'd suggest to add a `$force` parameter and skip that check when converting, and then add a maintenance tool that can clean out stale followers.

Reply:
I have to add a task that updates the user-profiles anyways! So I can combine both!

Reply:
I updated the `add_follower` function to be more error-tolerant. I just ignore HTTP responses with code 404 and 410 because they seem to be deleted.

Still missing: The update schedule, to finally remove profiles with errors (`errors >= 5` for example) and/or to updated profiles.